spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py

```diff
@@ -0,0 +1,754 @@
+            machine_time_step, time_scale_factor))
+        spec.switch_write_focus(
+            DataSpeedUpPacketGatherMachineVertex.DATA_REGIONS.CONFIG.value)
+
+        # the keys for the special cases
+        spec.write_value(base_key + 1)
+        spec.write_value(base_key + 2)
+
+        # locate the tag id for our data
+        spec.write_value(iptags[0].tag)
+
+        # End-of-Spec:
+        spec.end_specification()
+
+    @staticmethod
+    def _reserve_memory_regions(spec, system_size):
+        """ writes the dsg regions memory sizes. Static so that it can be used
+        by the application vertex.
+
+        :param spec: spec file
+        :param system_size: size of system region
+        :rtype: None
+        """
+
+        spec.reserve_memory_region(
+            region=DataSpeedUpPacketGatherMachineVertex.
+            DATA_REGIONS.SYSTEM.value,
+            size=system_size,
+            label='systemInfo')
+        spec.reserve_memory_region(
+            region=DataSpeedUpPacketGatherMachineVertex.DATA_REGIONS.
+            CONFIG.value,
+            size=DataSpeedUpPacketGatherMachineVertex.CONFIG_SIZE,
+            label="config")
+
+    @overrides(AbstractHasAssociatedBinary.get_binary_file_name)
+    def get_binary_file_name(self):
+        return self.static_get_binary_file_name()
+
+    @staticmethod
+    def static_get_binary_file_name():
+        """ used for supporting a application vertex that is used as \
+        only a data holder.
+
+        :return: aplex name
+        """
+        return "data_speed_up_packet_gatherer.aplx"
+
+    @overrides(AbstractProvidesLocalProvenanceData.get_local_provenance_data)
+    def get_local_provenance_data(self):
+        prov_items = list()
+        for (placement, memory_address, length_in_bytes) in \
+                self._provenance_data_items.keys():
+
+            # handle duplicates of the same calls
+            times_extracted_the_same_thing = 0
+            top_level_name = "Provenance_for_{}".format(self._label)
+            for (time_taken, lost_seq_nums) in self._provenance_data_items[
+                    (placement, memory_address, length_in_bytes)]:
+
+                # handle time
+                chip_name = "chip{}:{}".format(placement.x, placement.y)
+                last_name = "Memory_address:{}:Length_in_bytes:{}"\
+                    .format(memory_address, length_in_bytes)
+                iteration_name = "iteration{}".format(
+                    times_extracted_the_same_thing)
+                prov_items.append(ProvenanceDataItem(
+                    [top_level_name, "extraction_time", chip_name, last_name,
+                     iteration_name],
+                    time_taken, report=False, message=None))
+                times_extracted_the_same_thing += 1
+
+                # handle lost seq nums
+                lost_seq_num_iteration = 0
+                for n_lost_seq_nums in lost_seq_nums:
```

Comment:
Proper python style is to use `enumerate()` to make the counter for the collection being iterated over.

Reply:
fixed